pkg/syncutil/cachemanager/cachemanager.go

```diff
@@ -77,6 +228,139 @@ func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.
 	return nil
 }
 
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
 func (c *CacheManager) ReportSyncMetrics() {
 	c.syncMetricsCache.ReportSync()
 }
+
+func (c *CacheManager) syncGVKInstances(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	err := c.reader.List(ctx, u)
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	defer c.ReportSyncMetrics()
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+var gvksFailingTolist *watch.Set
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	// stopChan is used to stop any list operations still in progress
+	stopChan := make(chan bool, 1)
+	gvksFailingTolist = watch.NewSet()
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(stopChan)
+			return
+		case <-c.backgroundManagementTicker.C:
+			c.mu.Lock()
+			c.wipeCacheIfNeeded(ctx)
+
+			// spin up new goroutines to relist if new gvks to relist are
+			// populated from makeUpdates.
+			if c.gvksToList.Size() != 0 {
+				// stop any goroutines that were relisting before
+				// as we may no longer be interested in those gvks
+				stopChan <- true
+
+				// also try to catch any gvks that are in the aggregator
+				// but are failing to list from a previous replay.
+				for _, gvk := range gvksFailingTolist.Items() {
+					if c.specifiedGVKs.IsPresent(gvk) {
+						c.gvksToList.Add(gvk)
+					}
+				}
+
+				// save all gvks that need relisting
+				gvksToRelistForLoop := c.gvksToList.Items()
+
+				// clean state
+				gvksFailingTolist = watch.NewSet()
```

Comment:
`c.gvksToList` and `gvksFailingToList` seem like a more complicated way of answering the question "what needs to be listed"?

If you were to ask me what needs to be listed, I would say "anything that's specified that isn't in the cache".

Why not then make the logic reflect the domain?

We could have the following:

1. `CacheManager.cachedGVKs`, a `watch.Set` that tracks what GVKs have been added to the cache successfully.  This would mean they have been listed succesfully and added.  When the cache is wiped, this is set to empty.

2. `CacheManager.specifiedGVKs`.  We already have this.

3. When we decide what to pass to `replayLoop`, we just answer my question from above: `c.cachedGVKs.diff(c.specifiedGVKs)`.  We don't have to maintain any special variables as this method will generate a list of GVKs and pass it to `replayLoop` as a parameter.

In this scenario, we only add a GVK to `cachedGVKs` if it's successfully cached.  If there is a failure, we just leave that GVK out of our `cachedGVKs` and move on.  We'll try again later.

Reply:
I went w a `bool` to tell the `manageCache` loop to get all the aggregated gvks are relist them because of a subsequent `delete`